pkg/plugins/policies/meshpassthrough/plugin/xds/order.go

```diff
@@ -11,108 +11,135 @@ import (
 
 	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
 	api "github.com/kumahq/kuma/pkg/plugins/policies/meshpassthrough/api/v1alpha1"
+	"github.com/kumahq/kuma/pkg/util/pointer"
 )
 
 type MatchType int
 
 const (
-	Domain MatchType = iota + 1
-	WildcardDomain
-	IP
-	IPV6
+	WildcardDomain MatchType = iota + 1
+	Domain
 	CIDR
 	CIDRV6
+	IP
+	IPV6
 )
 
+var protocolOrder = map[core_mesh.Protocol]int{
+	core_mesh.ProtocolTLS:   0,
+	core_mesh.ProtocolTCP:   1,
+	core_mesh.ProtocolHTTP:  2,
+	core_mesh.ProtocolHTTP2: 2,
+	core_mesh.ProtocolGRPC:  2,
+}
+
 type Route struct {
 	Value     string
 	MatchType MatchType
 }
 
 type Matcher struct {
-	Protocol core_mesh.Protocol
-	Port     uint32
+	Protocol  core_mesh.Protocol
+	Port      uint32
+	MatchType MatchType
+	Value     string
 }
 
-type FilterChainMatcher struct {
-	Protocol core_mesh.Protocol
-	Port     uint32
-	Routes   []Route
+type FilterChainMatch struct {
+	Protocol  core_mesh.Protocol
+	Port      uint32
+	MatchType MatchType
+	Value     string
+	Routes    []Route
 }
 
-func GetOrderedMatchers(conf api.Conf) ([]FilterChainMatcher, error) {
-	matchers := map[Matcher]map[Route]bool{}
+func GetOrderedMatchers(conf api.Conf) ([]FilterChainMatch, error) {
+	matcherWithRoutes := map[Matcher]map[Route]bool{}
 	portProtocols := map[int]map[core_mesh.Protocol]bool{}
 	for _, match := range conf.AppendMatch {
-		var matchType MatchType
-		switch match.Type {
-		case api.MatchType("Domain"):
-			if strings.HasPrefix(match.Value, "*") {
-				matchType = WildcardDomain
-			} else {
-				matchType = Domain
-			}
-		case api.MatchType("IP"):
-			if govalidator.IsIPv6(match.Value) {
-				matchType = IPV6
-			} else {
-				matchType = IP
-			}
-		case api.MatchType("CIDR"):
-			split := strings.Split(match.Value, "/")
-			if govalidator.IsIPv6(split[0]) {
-				matchType = CIDRV6
-			} else {
-				matchType = CIDR
-			}
-		}
-		port := 0
-		if match.Port != nil {
-			port = *match.Port
-		}
+		port := pointer.DerefOr(match.Port, 0)
 		protocol := core_mesh.ParseProtocol(string(match.Protocol))
+		matchType := getMatchType(match)
+		matcher := Matcher{
+			Protocol:  protocol,
+			Port:      uint32(port),
+			MatchType: matchType,
```

Comment:
In the past, the `Matcher` only has 'Protocol' and 'Port' properties.
However, right now If we have the following policies, we would meet conflict because we would generate 2 elements in array `filterChainMatchers` (in the past we would only have one) :

```text
apiVersion: kuma.io/v1alpha1
kind: MeshPassthrough
metadata:
  name: one
  namespace: kuma-system
  labels:
    kuma.io/mesh: default
spec:
  targetRef:
    kind: Mesh
  default:
    passthroughMode: Matched
    appendMatch:
      - port: 80
        protocol: http
        type: Domain
        value: "www.google.com"
---
apiVersion: kuma.io/v1alpha1
kind: MeshPassthrough
metadata:
  name: two
  namespace: kuma-system
  labels:
    kuma.io/mesh: default
spec:
  targetRef:
    kind: MeshSubset
    proxyTypes:
      - Sidecar
    tags:
      kuma.io/service: nginx-service_kuma-demo_svc_80
  default:
    passthroughMode: Matched
    appendMatch:
      - port: 80
        protocol: http
        type: Domain
        value: "*.google.com"
```

Reply:
### old  orderedMatchers

```text
[{http 80 [{www.google.com 1} {*.google.com 2}]}]
```

### new orderedMatchers: 

```text
[{http 80 2  [{www.google.com 2}]} {http 80 1  [{*.google.com 1}]}]
```